### ISSUE TITLE
how to type X-expressions?

### ISSUE THREAD

**Issue by mbutterick** [racket/typed-racket#136]
Not a bug; more like a limitation. I have a lot of functions that operate on X-expressions and I’m unsure how to use them in TR without an `Xexpr` type. 

“So make one!” I’d like to, but an X-expression has this grammar:

```racket
  xexpr	 	=	 	string
 	 	|	 	(list symbol (list (list symbol string) ...) xexpr ...)
 	 	|	 	(cons symbol (list xexpr ...))
 	 	|	 	symbol
 	 	|	 	valid-char?
 	 	|	 	cdata
 	 	|	 	misc
```

Of course TR has `Symbol` and `String`. But in order to make an `Xexpr` type, I believe I’d also need primitive types for `Cdata` (which is a `String` with a special format), `Valid-Char` (which is an `Integer` in a certain range), and `Misc` (`String`s that are comments or processing instructions).

Or maybe there’s a workaround I’m overlooking.

**Comment by lexi-lambda** [comment]
This can actually be typed reasonably accurately with TR. Note that *cdata* and *misc* are structures, not raw strings, so these will need to be imported with `require/typed`. Doing that gets me this definition:

```racket
(require/typed
 xml
 [#:struct location ([line : (Option Natural)]
                     [char : (Option Natural)]
                     [offset : Natural])]
 [#:struct source ([start : (U location Symbol #f)]
                   [stop : (U location Symbol #f)])]
 [#:struct (cdata source) ([string : String])]
 [#:struct comment ([text : String])]
 [#:struct (p-i source) ([target-name : Symbol]
                         [instruction : String])])

(define-type X-Expr
  (U String
     (List* Symbol (Listof (List Symbol String)) (Listof X-Expr))
     (Pairof Symbol (Listof X-Expr))
     Symbol
     Natural
     cdata
     comment
     p-i))
```

The only piece that can't be precisely typed is `valid-char?`, so `Natural` suffices for now. I believe a closer type would be possible with refinement types, but those are currently only on @andmkent's fork.

**Comment by mbutterick** [comment]
That seems reasonable. Most `Naturals` are not `Valid-Chars`, however, so I still think that argues for a primitive type (cf. `Index`).

**Comment by samth** [comment]
I think, in most cases, it's not worth it to try to encode such precise invariants in the type, unless they really affect behavior (such as `sqrt`). Just having a typed x-expr will probably get almost all of the value here.

**Comment by mbutterick** [comment]
BTW, what is the advantage (in this case, but also generally) of using `require/typed` to import structs? By the time you declare all the fields and type them, you haven’t saved much work vs. just redefining the `struct` locally, e.g. compare this —

```racket
(require/typed
 xml
 [#:struct location ([line : (Option Natural)]
                     [char : (Option Natural)]
                     [offset : Natural])]
 [#:struct source ([start : (U location Symbol #f)]
                   [stop : (U location Symbol #f)])]
 [#:struct (cdata source) ([string : String])]
 [#:struct comment ([text : String])]
 [#:struct (p-i source) ([target-name : Symbol]
                         [instruction : String])])
```

with this, which AFAICT works equally well:

```racket
(define-struct location ([line : (Option Natural)]
                         [char : (Option Natural)]
                         [offset : Natural]))
(define-struct source ([start : (U location Symbol #f)]
                       [stop : (U location Symbol #f)]) #:transparent)
(define-struct comment ([text : String]) #:transparent)
(define-struct (p-i source) ([target-name : Symbol]
                             [instruction : String]) #:transparent)
(define-struct (pcdata source) ([string : String]) #:transparent)
(define-struct (cdata source) ([string : String]) #:transparent)
```

“Well, you lose the connection to the underlying `xml` library.” Right, but `require/typed` is already sensitive to implementation details. So if you have to track those details anyhow, why take on the overhead of `require/typed`?

**Comment by lexi-lambda** [comment]
Because those structs aren't the same as the structs the XML library uses, so if you try to call `cdata-string` from the typed side on a `cdata` instance created on the untyped side, it will fail with a contract error. You'd actually get a contract error even earlier, since TR would catch the mismatched structs as soon as the data came in from the untyped side.

You can see that illustrated with this small program.

```racket
#lang racket

(module u racket
  (struct point (x y))
  (provide make-origin)
  (define (make-origin)
    (point 0 0)))

(module t typed/racket
  (struct point ([x : Real] [y : Real]))
  (require/typed
   (submod ".." u)
   [make-origin (-> point)])
  (make-origin))

(require 't)
```

Basically, structs with the same name aren't the same type of struct, they're still disparate types.
